File: room.js
'use strict';
const util = require('util');

const [ colorEnum ] = util.enumToMap([
	COLOR_BLUE,
	COLOR_BROWN,
	COLOR_CYAN,
	COLOR_GREEN,
	COLOR_GREY,
	COLOR_ORANGE,
	COLOR_PURPLE,
	COLOR_RED,
	COLOR_WHITE,
	COLOR_YELLOW,
]);

let flagLayout;
let mineralLayout;
let sourceLayout;
let resourceLayout;
let tombstoneLayout;

module.exports = {
	init(layouts) {
		flagLayout = layouts.flag;
		mineralLayout = layouts.mineral;
		sourceLayout = layouts.source;
		resourceLayout = layouts.resource;
		tombstoneLayout = layouts.tombstone;
	},

	writeMineral(env, ptr, mineral) {
		if (mineral === undefined) {
			env.HEAPU32[(ptr + mineralLayout.type) >> 2] = 0;
		} else {
			env.HEAPU32[(ptr + mineralLayout.type) >> 2] = mineral.mineralType;
			env.HEAPU32[(ptr + mineralLayout.amount) >> 2] = mineral.mineralAmount;
			env.HEAPU32[(ptr + mineralLayout.density) >> 2] = mineral.density;
			env.HEAPU32[(ptr + mineralLayout.ticksToRegeneration) >> 2] = mineral.ticksToRegeneration;
		}
	},
};

File: room.cc
#include "./room.h"
#include <emscripten.h>

void room_t::init() {
	EM_ASM({
		Module.screeps.room.init({
			'flag': {
				'sizeof': $0,
				'color': $1,
				'secondaryColor': $2,
			},
			'mineral': {
				'size': $3,
				'type': $4,
				'amount': $5,
				'density': $6,
				'ticksToRegeneration': $7,
			},
			'source': {
				'sizeof': $8,
				'energy': $9,
				'energyCapacity': $10,
				'ticksToRegeneration': $11,
			},
			'resource': {
				'sizeof': $12,
				'resourceType': $13,
				'amount': $14,
			},
			'tombstone': {
				'sizeof': $15,
				'creep': $16,
				'deathTime': $17,
				'store': $18,
				'ticksToDecay': $19,
			},
		});
	},
		sizeof(flag_t),
		offsetof(flag_t, color),
		offsetof(flag_t, secondary_color),
		sizeof(mineral_t),
		offsetof(mineral_t, type),
		offsetof(mineral_t, amount),
		offsetof(mineral_t, density),
		offsetof(mineral_t, ticks_to_regeneration),
		sizeof(source_t),
		offsetof(source_t, energy),
		offsetof(source_t, energy_capacity),
		offsetof(source_t, ticks_to_regeneration),
		sizeof(dropped_resource_t),
		offsetof(dropped_resource_t, type),
		offsetof(dropped_resource_t, amount),
		sizeof(tombstone_t),
		offsetof(tombstone_t, creep),
		offsetof(tombstone_t, death_time),
		offsetof(tombstone_t, store),
		offsetof(tombstone_t, ticks_to_decay)
	);
}
